ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,164 @@
+// Libraries
+import {sortBy} from 'lodash'
+
+// Utils
+import {asAssignment, getAllVariables} from 'src/variables/selectors'
+import {buildVarsOption} from 'src/variables/utils/buildVarsOption'
+import {filterUnusedVarsBasedOnQuery} from 'src/shared/utils/filterUnusedVars'
+import {runQuery} from 'src/shared/apis/query'
+
+// Types
+import {RunQueryResult} from 'src/shared/apis/query'
+import {CancelBox} from 'src/types/promises'
+import {AppState, Variable} from 'src/types'
+
+export const TIME_INVALIDATION = 5000
+
+const asSimplyKeyValueVariables = (vari: Variable) => {
+  return {
+    [vari.name]: vari.selected || [],
+  }
+}
+
+// Hashing function found here:
+// https://jsperf.com/hashcodelordvlad
+// Through this thread:
+// https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
+const hashCode = (queryText: string): string => {
+  let hash = 0,
+    char
+  if (!queryText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < queryText.length; i++) {
+    char = queryText.charCodeAt(i)
+    hash = (hash << 5) - hash + char
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  _cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (id: string, variableID: string): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].variables !== variableID) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    return this.cache[id].values
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    variableID: string,
+    values: RunQueryResult,
+    isCustomTime: boolean = false
+  ): void => {
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      isCustomTime,
+      values,
+      variables: variableID,
```

Comment:
why is this called `variables` when it's a `variableID`?

Reply:
The variableID is the hashed version of the variables used

Reply:
that's really interesting! is there any way to indicate that `variables` are the hash of variables? documenting that in variable name would be very helpful to a reader!

Reply:
Definitely. How would you feel about `hashedVariables` as a property & parameter name rather than `variables` and `variableID` (respectively)?